src/vmm/tests/integration_tests.rs

```diff
@@ -362,13 +370,61 @@ fn create_snapshot(is_diff: bool) {
             assert_eq!(restored_microvm_state.vcpu_states.len(), 1);
         }
     }
+    (snapshot_file, memory_file)
+}
+
+#[cfg(target_arch = "x86_64")]
+fn verify_load_snapshot(snapshot_file: TempFile, memory_file: TempFile) {
+    use vm_memory::GuestMemoryMmap;
+    use vmm::memory_dump::SnapshotMemory;
+
+    let pid = unsafe { libc::fork() };
+    match pid {
+        0 => {
+            set_panic_hook();
+            let mut event_manager = EventManager::new().unwrap();
+            let empty_seccomp_filter = get_seccomp_filter(SeccompLevel::None).unwrap();
+
+            snapshot_file.as_file().seek(SeekFrom::Start(0)).unwrap();
+            let microvm_state: MicrovmState =
+                Snapshot::load(&mut snapshot_file.as_file(), VERSION_MAP.clone()).unwrap();
+            let mem = GuestMemoryMmap::restore(memory_file.as_file(), &microvm_state.memory_state)
+                .unwrap();
+
+            let vmm = build_microvm_from_snapshot(
+                &mut event_manager,
+                microvm_state,
+                mem,
+                false,
+                &empty_seccomp_filter,
+            )
+            .unwrap();
+            // For now we're happy we got this far, we don't test what the guest is actually doing.
+            vmm.lock().unwrap().stop(0);
+        }
+        vmm_pid => {
+            // Parent process: wait for the vmm to exit.
+            let mut vmm_status: i32 = -1;
+            let pid_done = unsafe { libc::waitpid(vmm_pid, &mut vmm_status, 0) };
+            assert_eq!(pid_done, vmm_pid);
+            restore_stdin();
+            // If any panics occurred, its exit status will be != 0.
+            assert!(unsafe { libc::WIFEXITED(vmm_status) });
+            assert_eq!(unsafe { libc::WEXITSTATUS(vmm_status) }, 0);
+        }
+    }
 }
 
 #[cfg(target_arch = "x86_64")]
 #[test]
-fn test_create_snapshot() {
+fn test_create_and_load_snapshot() {
     // Create diff snapshot.
-    create_snapshot(true);
+    let (snapshot_file, memory_file) = verify_create_snapshot(true);
+    // Create a new microVm from snapshot.
+    verify_load_snapshot(snapshot_file, memory_file);
```

Comment:
Are we testing here loading from a diff snapshot?

Reply:
The problem appears after we resume the microVM, it should encounter no error before. The test seems to be about verifying that `create_snapshot` produces valid snapshot and memory files and that `load_snapshot` can build the microVM from these files. The actual content of the memory file and what happens after resume can be checked in another test, I think. To test functionality after loading from a diff snapshot we'd also need a tool for merging the sparse files.

Reply:
Yup, it's kinda confusing though to have a test that is basically checking for a happy case when that case is expected to not be a happy one :-?, the microVM can be built, but it can't be used. I don't know if it makes much sense to test loading a microVM from snapshot without resuming it too.
Anyway, a comment here why is this working (since we're telling in docs that loading should be done only from a full snapshot) would be useful.

Reply:
Most of our rust integration tests only verify functions run with no errors but don't verify the guest workload. I think the main reason for that is that we have our python integration tests which have a complex framework for doing that.

To verify how guest behaves in rust integration tests would need framework to communicate with the guest; to verify devices still function in the guest would need framework to set up devices on the host and framework to test black-box device functionality; etc.

So the current test only verifies that *building* a microvm works from a *code-logic* point of view.

Reply:
Got it, still, a more detailed comment on line 423 would be great.

Reply:
done